blog post to ask for community help about biased locking (#747)
quarkusio/quarkusio.github.io: _posts/2020-09-28-biased-locking-help.adoc

---
layout: post
title: Biased locking removed from Java - does it affect you ?
date: 2020-09-28
tags: help

> synopsis: In Java 15 *biased locking* is removed from the Java Virtual Machine as per JEP 374. We need your help seeing if it affects Quarkus performance in your usecase.

Reviewer: AIUI in JDK 15 it is no longer the *default* in JDK 15 and might be removed in 16 or 17.

Reviewer 2: Yes in jdk15 the default has been reset to disabled. It can still be enabled. This should say something like

`In Java 15 *biased locking* has been deprecrated as per JEP 374. This means the default setting has been switched from enabled to disabled and the plan is to remove support for it from the Java Virtual Machine in a pending release.`

Author: ive pushed a new version - let me know if that works for you.

> synopsis: In Java 15 *biased locking* is removed from the Java Virtual Machine as per JEP 374. We need your help seeing if it affects Quarkus performance in your usecase.

Author: ```suggestion
synopsis: In Java 15 *biased locking* is now disabled by default and is planned to be removed in the Java Virtual Machine as per JEP 374. We need your help seeing if it affects Quarkus performance in your usecase.
```

author: maxandersen
---

> Last week we became aware that the OpenJDK team in Java 15 have disabled so called *biased locking* in the Java virtual machine. This is change from previous years and could *potentially* have a negative impact on Java applications performance.

Reviewer: s/applications/application
or s/Java applications/a Java application's/g

> Last week we became aware that the OpenJDK team in Java 15 have disabled so called *biased locking* in the Java virtual machine. This is change from previous years and could *potentially* have a negative impact on Java applications performance.

Author: ```suggestion
Last week we became aware that the OpenJDK team in Java 15 have disabled so called *biased locking* in the Java virtual machine. This is change from previous years and could *potentially* have a negative impact on Java application's performance.
```

Red Hat's own performance teams are currently running performance tests to see how it affects our Java middleware, but no amount of generic testing can reveal how this change affects real world usage.

That is where you come in.

> We would like to get info from you on wether your application performance is affected by biased locking or not.

Reviewer 2: wether should be whether

Author: ```suggestion
We would like to get info from you on whether your application performance is affected by biased locking or not.
```

To do so please try the following in your application performance tests:

> Run your Qukarus application tests/performance as you normally would with the following command line flags on Java 11 (jdk11u) and if feasible Java 15 (jdk15):

Reviewer: s/Qukarus/Quarkus

Author: ```suggestion
Run your Quarkus application tests/performance as you normally would with the following command line flags on Java 11 (jdk11u) and if feasible Java 15 (jdk15):
```

*enabled*: -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0

*disabled*: -XX:-UseBiasedLocking

We would like to know the result of this no matter if you see a regression or not on the same Java Virtual Machine.

If you are able then running it with the following scenarios would be of great help too:

 . single-threaded
 . thread count ~= hardware core count

>  . thread count ~= N * hardware core count where 8 < N < 16

Reviewer: Is this section meaningful for a Quarkus app? For WildFly I planned to omit this kind of discussion.

Reviewer 2: Well, if a Quarkus app or performance regression test can be run with a  user-configured number of threads then yes. That may not be an option of course.

Author: ah - good point; I was struggling actually wording it correctly. lets just keep it simple and ask for delta effect of running with or without biased locking enabled.

Reviewer 2: I understand this is not a very clear recommendation for many users. However, I don' think it is a good idea for us to simply to ignore how each app uses threads and how the thread count relates to the underlying hw thread capacity. Leaving that out of the question we are asking may well make it harder for us to assess the real nature of the problem.

Of course, for many apps the number of threads that will run in a deployment  will be constant. But in others it may be a configurable parameter and that's something we really need to factor in. If there is room for variation of the number of  threads that an app might use then there is immense room for both contedned and uncontended locking by those threads to influence outcome of enabling or disabling biased locking. In such cases a single data point will often be very misleading (especially if the only data point is 1 thread).

Author: i've pushed a version with a little bit context and make clear its only if they are able to.

trying to avoid it being "hard" to send feedback.

=== Reporting the result

Please open a bug on with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:

```
jvm used: jdk11 or jdk15
biased locking: enabled or disabled
thread-count: 1..N

> hardward-cores: N

Author: ```suggestion
hardware-core count: N
```

```

This will help us immensly. Thank you!

== Background

> Below are some background context on biased locking - all optional reading - you don't need to understand the details to help us by running your performance tests and let us know if anything changes.

Reviewer 2: Below /is/ ...

> Below are some background context on biased locking - all optional reading - you don't need to understand the details to help us by running your performance tests and let us know if anything changes.

Author: ```suggestion
Below is some background context on biased locking - all optional reading - you don't need to understand the details to help us by running your performance tests and let us know if anything changes.
```

=== What is biased locking?

Biased locking lowers the cost of /uncontended/ synchronization.

*Without biased locking*: a thread needs to set and clear a lock bit
when it performs repeated synchronizations on the same object. It also needs to wait for those set/clear writes to be drained to local cache before proceeding to execute further memory operations.

> *With biased locking*: the first time a thread synchronizes on an object it does a bit more work to acquire ownership of the synchronizated object ('bias' it to the thread). Subsequent syncrhonizations proceed via a simple read test with no need to  drain to cache.

Reviewer 2: `ownership of the synchronizated object` should be synchronized

> *With biased locking*: the first time a thread synchronizes on an object it does a bit more work to acquire ownership of the synchronizated object ('bias' it to the thread). Subsequent syncrhonizations proceed via a simple read test with no need to  drain to cache.

Author: ```suggestion
*With biased locking*: the first time a thread synchronizes on an object it does a bit more work to acquire synchronized ('bias' it to the thread). Subsequent syncrhonizations proceed via a simple read test with no need to  drain to cache.
```

Where's the trade off? Well, if a biased lock is contended then
there is more work to do to bias and unbias the lock. However, it is known that many synchronized operations are uncontended.

Additionally, so called 'lock-free' code (i.e. code that uses atomics or `cmpxchg` instead of 'synchronized' operations) bypasses these concerns.

Biasing can be big win when a potentially concurrent data structure is actually used sequentially. The case where it helps most is exemplified in the problem we already found in class `DataOutputStream`. Normally only
one thread writes a `DataOutputStream` and it is often not read until the stream has been filled. All the same, every `putInt()` or `putLong()`` call invokes a syncrhonized method to increment the byte count by 4 or 8.
That's needed just in case some other thread might want to reliably
locate the end of the valid buffer data but that rarely happens. So, the unbiased case suffers lock write and cache drain delays at every basic put operation.

> 

Reviewer 2: Add a new, much worse example here.

    A similar case occurs with class ByteOutput Stream. Method putByte is synchronized. So writing a single byte involves a lock and unlock. n.b. method putInt calls putByte 4 times, requiring 4 locks and unlocks. Method putLong calls it 8 times!

> locate the end of the valid buffer data but that rarely happens. So, the unbiased case suffers lock write and cache drain delays at every basic put operation.

Author: ```suggestion
locate the end of the valid buffer data but that rarely happens. So, the unbiased case suffers lock write and cache drain delays at every basic put operation.

A similar case occurs with class ByteOutput Stream. Method putByte is synchronized. So writing a single byte involves a lock and unlock. n.b. method putInt calls putByte 4 times, requiring 4 locks and unlocks. Method putLong calls it 8 times!
```

== Why is biased locking being removed?

The implementation of biased locking adds a great deal of complexity to the JVM and is understood by only a small subset of the most experienced engineers. The cost of maintaining it and designing around it is significantly slowing down progress on new features. It has been a long term goal to remove it if at all possible. Some OpenJDK contributors wanted to remove it

> right away in jdk15 others argues for a slower deprecation toute in order to check that we could really dispense with it.

Reviewer 2: `deprecation toute` should be `deprecation route`

Reviewer: s/argues/argue

> right away in jdk15 others argues for a slower deprecation toute in order to check that we could really dispense with it.

Author: ```suggestion
right away in jdk15 others argue for a slower deprecation route in order to check that we could really dispense with it.
```

== What happens next?

> We are collecting our own internal performance tests across multiple teams in Red Hat and will gather data from community reported tests too and see what the data indicates. At this stage we don't know what we don't know.

Reviewer 2: `At this stage we don't know what we don't know.` rep[lace this with

`At this stage we are making no assumption that the removal of biased locking will definitely make performance worse. We know that in some cases not having biased locks will improve performance. Our concern is to find cases, like the JDK examples above, where it might cause serious performance degradation and get an idea of how bad, and also how common, the worst cases might be.`

> We are collecting our own internal performance tests across multiple teams in Red Hat and will gather data from community reported tests too and see what the data indicates. At this stage we don't know what we don't know.

Author: ```suggestion
We are collecting our own internal performance tests across multiple teams in Red Hat and will gather data from community reported tests too and see what the data indicates. At this stage we are making no assumption that the removal of biased locking will definitely make performance worse. We know that in some cases not having biased locks will improve performance. Our concern is to find cases, like the JDK examples above, where it might cause serious performance degradation and get an idea of how bad, and also how common, the worst cases might be.
```

Red Hat's own performance teams are currently running performance tests to see how it affects our Java middleware, but no amount of generic testing can reveal how this change affects real world usage.

That is where you come in.

We would like to get info from you on whether your application performance is affected by biased locking or not.

To do so please try the following in your application performance tests:

Run your Quarkus application tests/performance as you normally would with the following command line flags on Java 11 (jdk11u) and if feasible Java 15 (jdk15):

*enabled*: -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0

*disabled*: -XX:-UseBiasedLocking

We would like to know the result of this no matter if you see a regression or not on the same Java Virtual Machine.

If you are able then running it with the following scenarios would be of great help too:

 . single-threaded
 . thread count ~= hardware core count
 . thread count ~= N * hardware core count where 8 < N < 16

 The intent of these are to see how the level of concurrency affects the result.

=== Reporting the result

> Please open a bug on with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:

Reviewer: Is there a missing link after "open a bug on" ?

Author: ```suggestion
Please [open a bug](https://github.com/quarkusio/quarkus/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.md&title=[jep374]) with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:
```

Author: ```suggestion
Please [open a bug](https://github.com/quarkusio/quarkus/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.md&title=%28jep374%29) with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:
```

> Please open a bug on with `[jep374]` in the title + your project. i.e. `[jep374] results for acme project crazy panda` containing the following information per run in the description:

Author: this is fixed in latest

```
jvm used: jdk11 or jdk15
biased locking: enabled or disabled
thread-count: 1..N (if you know)

> hardware-core count: N (if you know)

Reviewer 3: In order to quantify any impact, we need to see the results with and without biased locking enabled

```suggestion
hardware-core count: N (if you know)
performance test result: with biased locking
performance test result: without biased locking
```

```

This will help us immensly. Thank you!

== Background

Below is some background context on biased locking - all optional reading - you don't need to understand the details to help us by running your performance tests and let us know if anything changes.

=== What is biased locking?

Biased locking lowers the cost of /uncontended/ synchronization.

*Without biased locking*: a thread needs to set and clear a lock bit
when it performs repeated synchronizations on the same object. It also needs to wait for those set/clear writes to be drained to local cache before proceeding to execute further memory operations.

*With biased locking*: the first time a thread synchronizes on an object it does a bit more work to acquire synchronized ('bias' it to the thread). Subsequent syncrhonizations proceed via a simple read test with no need to  drain to cache.

Where's the trade off? Well, if a biased lock is contended then
there is more work to do to bias and unbias the lock. However, it is known that many synchronized operations are uncontended.

> Additionally, so called 'lock-free' code (i.e. code that uses atomics or `cmpxchg` instead of 'synchronized' operations) bypasses these concerns.

Reviewer 2: This sentence is incorrect and should just be removed:

`Additionally, so called 'lock-free' code ... `

Author: ```suggestion
```
